### PR TITLE
Types from header

### PR DESCRIPTION
**Pull request by PeterNSteinmetz**
Factors out parsing of information from header of file. Add determination of recording type from header information. 

Changes indexes to indices in argument names again -- will need determination of acceptability of this spelling change for files being worked on only. See https://grammarist.com/usage/indexes-indices/ for discussion of usage, particularly in technical contexts. 

This factorization is in preparation for larger changes to accommodate early file styles with different header structures. Provided here as a smaller chunk in order to avoid very much larger changes which are harder to check and understand.

### CODE REVIEW

**Comment by apdavison** [comment]
Many thanks for this. A quick comment on the "indices/indexes" question. Having grown up in the UK, I prefer "indices" in general. However, we decided to use "indexes" in Neo because it is less confusing for non-native English speakers and more familiar to North Americans. It's fine to use "indices" for internal variables, but any arguments that are exposed to Neo users as part of the API should be "indexes".

**Comment by PeterNSteinmetz** [comment]
> A quick comment on the "indices/indexes" question. Having grown up in the UK, I prefer "indices" in general. However, we decided to use "indexes" in Neo because it is less confusing for non-native English speakers and more familiar to North Americans. It's fine to use "indices" for internal variables, but any arguments that are exposed to Neo users as part of the API should be "indexes".

So how about the case of the argument to _get_analogsignal_chunk ? That is nominally a private method but I imagine a lot of people use that. Should those also be 'indexes' ?

**Comment by PeterNSteinmetz** [comment]
circleci is complaining that the test is failing because trying to download the file https://web.gin.g-node.org/NeuralEnsemble/ephy_testing_data/raw/master/neuralynx/Cheetah_v4.0.2/original_data/CSC14_trunc.Ncs is returning a 404 error. 

Yet that file is available and downloadable in a browser or if I pull it using git from the gin-g repository. 

What is going wrong there?

**Comment by JuliaSprenger** [comment]
Hi @PeterNSteinmetz, I restarted the tests and the files could now be successfully downloaded from gin. It seems this was a temporary issue with gin. The tests are succeeding now.

**Comment by JuliaSprenger** [comment]
> > A quick comment on the "indices/indexes" question. Having grown up in the UK, I prefer "indices" in general. However, we decided to use "indexes" in Neo because it is less confusing for non-native English speakers and more familiar to North Americans. It's fine to use "indices" for internal variables, but any arguments that are exposed to Neo users as part of the API should be "indexes".
> 
> So how about the case of the argument to _get_analogsignal_chunk ? That is nominally a private method but I imagine a lot of people use that. Should those also be 'indexes' ?

Yes, I think we should stay consistently with `channel_indexes` for now and discuss a potential name change in a separate issue. Could you please also use `indexes` for `_get_analogsignal_chunk`?

**Comment by PeterNSteinmetz** [comment]
I will merge my NlxFixes branch into this PR, the TypesFromHeader branch, thereby fixing the channel_indexes question, once PR #851 is merged onto main. I want to leave NlxFixes open for now until that happens.

**Comment by JuliaSprenger** [comment]
#851 is now merged in master and here. Are you planning to address the todos introduced in this PR also here or is this PR ready for review / merge from your side?

**Comment by PeterNSteinmetz** [comment]
That is it for this branch. Please merge into master.

**Comment by PeterNSteinmetz** [comment]
This is ready for merging.

File: neo/rawio/neuralynxrawio.py [hunk 1]
```diff
@@ -531,171 +588,276 @@ def read_ncs_files(self, ncs_filenames):
...
+
+        # close time, if available
+        if 'datetime2_regex' in hpd:
+            dt2 = re.search(hpd['datetime2_regex'], txt_header).groupdict()
+            info['recording_closed'] = datetime.datetime.strptime(
+                dt2['date'] + ' ' + dt2['time'], hpd['datetimeformat'])
+
+        return info
+
+    def typeOfRecording(self):
```
Comment (JuliaSprenger) [inline neo/rawio/neuralynxrawio.py:813]: This function name should be lower case and with underscores for compatibility with PEP8 and consistency with other functions in Neo.

File: neo/rawio/neuralynxrawio.py [hunk 2]
```diff
@@ -87,15 +97,18 @@ def _parse_header(self):
...
-            if (os.path.getsize(filename) <= HEADER_SIZE) and (ext in ['ncs']):
+            # Skip Ncs files with only header. Other empty file types
+            # will have an empty dataset constructed later.
+            if (os.path.getsize(filename) <= NlxHeader.HEADER_SIZE) and ext in ['ncs']:
                 self._empty_ncs.append(filename)
                 continue
 
             # All file have more or less the same header structure
-            info = read_txt_header(filename)
+            info = NlxHeader.buildForFile(filename)
```
Comment (JuliaSprenger) [inline neo/rawio/neuralynxrawio.py:111]: This function name should be lower case and with underscores for compatibility with PEP8 and consistency with other functions in Neo.

**Review by JuliaSprenger — APPROVED** [review]
There are a few naming issues I found, but otherwise the PR looks ready to merge to me. Thanks!

File: neo/rawio/neuralynxrawio.py (hunk 2 above)
Comment (PeterNSteinmetz) [reply neo/rawio/neuralynxrawio.py]: Those changes incorporated onto this branch.

**Review by JuliaSprenger — APPROVED** [review]

**Comment by PeterNSteinmetz** [comment]
I think tests are still failing on hdf5io .

**Comment by PeterNSteinmetz** [comment]
OK, just merged NeuralEnsemble:master back into my TypeFromHeader branch. Let's see how this goes.

**Pull request merged**